### PR TITLE
feat: add rolling restart feature

### LINKED ISSUE #288: Rolling restart
To do a rolling restart, we must add a pause between component restarts. The user must be able to specify the time of this pause.

### PR DESCRIPTION
**Pull request by rpignolet**
Add `--rolling-interval` option for `tdp plan reconfigure` command.

#### Which issue(s) this PR fixes

Fixes #288 

#### Agreements

- [x] I hereby declare this contribution to be licensed under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0).
- [x] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

**Review by PaulFarault — COMMENTED** [review]
Should we create a method to check if the `wait_sleep` operation exist in the collection at init time?

File: tdp/core/collections.py [hunk 1]
```diff
@@ -242,6 +245,14 @@ def get_operation(self, operation_name: str) -> Operation:
             )
         return self.operations[operation_name]
 
+    def get_sleep_operation(self) -> Tuple[Operation, str]:
```
[the remark this answers is not in the document]
Comment (rpignolet) [reply tdp/core/collections.py]: This method is removed.

File: tdp/core/models/deployment_log.py [hunk 2]
```diff
@@ -275,19 +280,64 @@ def from_stale_components(
...
+                            operation=operation.name,
+                            operation_order=operation_order,
+                            host=host,
+                            extra_vars=None,
+                            state=OperationStateEnum.PLANNED,
+                        )
+                    )
+                    # Add sleep operation between each "restart"
+                    # and do not add a sleep operation after the last "restart".
+                    if operation.action_name == "restart" and i < len(hosts) - 1:
```
[the remark this answers is not in the document]
Comment (rpignolet) [reply tdp/core/models/deployment_log.py]: Done.

File: tdp/cli/commands/plan/reconfigure.py [hunk 3]
```diff
@@ -10,17 +10,24 @@
 
 
 @click.command(short_help="Restart required TDP services.")
+@click.option(
+    "--rolling-interval",
+    envvar="TDP_ROLLING_INTERVAL",
+    type=int,
+    help="Enable rolling restart with specific waiting time between component restart.",
```
[the remark this answers is not in the document]
Comment (rpignolet) [reply tdp/cli/commands/plan/reconfigure.py]: Done.

**Comment by rpignolet** [comment]
> Should we create a method to check if the `wait_sleep` operation exist in the collection at init time?

Yes I also wondered the same thing. If the user does not use the rolling functionality, I don't see why we should make an error if the sleep operation does not exist. Normally an error will be raised at `tdp plan reconfigure` time using the `--rolling-interval` option if the sleep operation does not exist.

File: tdp/cli/commands/plan/reconfigure.py [hunk 4]
```diff
@@ -10,17 +10,24 @@
 
 
 @click.command(short_help="Restart required TDP services.")
+@click.option(
+    "--rolling-interval",
```
Comment (PaulFarault) [inline tdp/cli/commands/plan/reconfigure.py:14]: should we add a `-ri` alternative?

File: tdp/core/models/deployment_log.py [hunk 5]
```diff
@@ -80,6 +81,9 @@ class DeploymentLog(Base):
     extra_vars: Mapped[Optional[list[str]]] = mapped_column(
         JSON(none_as_null=True), doc="List of extra vars."
     )
+    rolling_interval: Mapped[Optional[int]] = mapped_column(
+        doc="Deployment rolling interval."
```
Comment (PaulFarault) [inline tdp/core/models/deployment_log.py:85]:
```suggestion
        doc="Deployment rolling interval (in sec)."
```

File: tdp/core/models/deployment_log.py [hunk 6]
```diff
@@ -285,19 +294,35 @@ def from_stale_components(
...
         )
-        deployment_log.operations = [
-            OperationLog(
-                operation=operation.name,
-                operation_order=i,
-                host=host,
-                extra_vars=None,
-                state=OperationStateEnum.PLANNED,
+        operation_order = 1
+        for operation, host in operation_hosts:
```
Comment (PaulFarault) [inline tdp/core/models/deployment_log.py:302]: Can't we use enumerate here as there is no nested logic anymore?

File: tdp/core/models/deployment_log.py [hunk 7]
```diff
@@ -285,19 +294,35 @@ def from_stale_components(
...
+                    state=OperationStateEnum.PLANNED,
+                )
             )
-            for i, (operation, host) in enumerate(operation_hosts, 1)
-        ]
+            # Add sleep operation after each "restart"
+            if rolling_interval is not None and operation.action_name == "restart":
+                operation_order += 1
+                deployment_log.operations.append(
+                    OperationLog(
```
Comment (PaulFarault) [inline tdp/core/models/deployment_log.py:316]: Shouldn't we add a default log message for waiting_operations?

File: tdp/cli/commands/plan/reconfigure.py (hunk 4 above)
Comment (rpignolet) [reply tdp/cli/commands/plan/reconfigure.py:15]: Done.

File: tdp/core/models/deployment_log.py (hunk 5 above)
Comment (rpignolet) [reply tdp/core/models/deployment_log.py]: Done.

File: tdp/core/models/deployment_log.py (hunk 6 above)
Comment (rpignolet) [reply tdp/core/models/deployment_log.py:302]: There is no nested logic.

**Review by PaulFarault — APPROVED** [review]

**Review by mdrutel — APPROVED** [review]
tested OK

**Pull request merged**